File: packages/octospaces-sdk/src/spaces/registry.ts
/**
 * Space registries (plaintext metadata docs). A user's spaces live at
 * `user/<userId>/_spaces`; each space's ACCESS RECORD (owner/members + shared
 * name/image) at `spaces/<spaceId>/_access`. The object tree lives in the encrypted
 * unified object index (`objects/_index`, see `object-index.ts`); `_access` is the
 * owner-only access record.
 */
import { ConflictError, StarfishHttpError } from '@drakkar.software/starfish-client';
import type { StarfishClient } from '@drakkar.software/starfish-client';

import type { ArchivedDms, CapMap, DmMap, MutePrefs, PubAccessMap, ReadPrefs, Space, SpaceVisibility } from '../core/types.js';
import type { SealedBlob } from '../sync/account-seal.js';
import { randomId } from '../core/ids.js';
import type { Session } from '../sync/identity.js';
import { seedSpaceObjectIndex } from './object-index.js';
import { spaceAccessPull, spaceAccessPush, spacesPull, spacesPush } from '../sync/paths.js';

/** Owner-set, SHARED space identity, persisted in the `_access` registry doc
 *  (plaintext — NOT E2EE). `image` is a data URI. Both optional for back-compat. */
export interface SpaceMeta {
  name?: string | null;
  image?: string | null;
  visibility?: SpaceVisibility;
}

/** A resolved name/image update fanned out so the SpacesProvider adopts a
 *  freshly-reconciled value without waiting for its next navigation refresh. */
export interface SpaceMetaUpdate {
  name: string;
  short: string;
  image?: string;
}

const spaceMetaListeners = new Set<(spaceId: string, meta: SpaceMetaUpdate) => void>();

export function onSpaceMeta(fn: (spaceId: string, meta: SpaceMetaUpdate) => void): () => void {
  spaceMetaListeners.add(fn);
  return () => { spaceMetaListeners.delete(fn); };
}

export function broadcastSpaceMeta(spaceId: string, meta: SpaceMetaUpdate): void {
  for (const fn of spaceMetaListeners) fn(spaceId, meta);
}

interface SpacesDoc {
  spaces: Space[];
  caps: CapMap;
  mutes: MutePrefs;
  reads: ReadPrefs;
  pubAccess: PubAccessMap;
  dms: DmMap;
  quickReactions: string[];
  archivedDms: ArchivedDms;
  hash: string | null;
}

function coerceDms(raw: unknown): DmMap {
  const src = raw && typeof raw === 'object' ? (raw as Record<string, unknown>) : {};
  const out: DmMap = {};
  for (const [k, v] of Object.entries(src)) if (typeof v === 'string') out[k] = v;
  return out;
}

function coerceMutes(raw: unknown): MutePrefs {
  const r = (raw && typeof raw === 'object' ? raw : {}) as { rooms?: unknown; spaces?: unknown };
  const pick = (v: unknown): Record<string, true | number> =>
    v && typeof v === 'object' ? (v as Record<string, true | number>) : {};
  return { rooms: pick(r.rooms), spaces: pick(r.spaces) };
}

function coerceReads(raw: unknown): ReadPrefs {
  const r = (raw && typeof raw === 'object' ? raw : {}) as { rooms?: unknown };
  const src = r.rooms && typeof r.rooms === 'object' ? (r.rooms as Record<string, unknown>) : {};
  const rooms: Record<string, number> = {};
  for (const [id, v] of Object.entries(src)) if (typeof v === 'number' && Number.isFinite(v)) rooms[id] = v;
  return { rooms };
}

function coerceQuickReactions(raw: unknown): string[] {
  return Array.isArray(raw) ? raw.filter((v): v is string => typeof v === 'string') : [];
}

function coerceArchivedDms(raw: unknown): ArchivedDms {
  const src = raw && typeof raw === 'object' ? (raw as Record<string, unknown>) : {};
  const out: ArchivedDms = {};
  for (const [k, v] of Object.entries(src)) if (v === true) out[k] = true;
  return out;
}

async function pullSpacesDoc(client: StarfishClient, userId: string): Promise<SpacesDoc> {
  const res = await client.pull(spacesPull(userId)).catch((err: unknown) => {
    if (err instanceof StarfishHttpError && err.status === 404) return null;
    throw err;
  });
  const data = res?.data as
    | {
        spaces?: Space[];
        caps?: CapMap;
        mutes?: unknown;
        reads?: unknown;
        pubAccess?: PubAccessMap;
        dms?: unknown;
        quickReactions?: unknown;
        archivedDms?: unknown;
      }
    | undefined;
  return {
    spaces: Array.isArray(data?.spaces) ? data!.spaces! : [],
    caps: data?.caps && typeof data.caps === 'object' ? data.caps : {},
    mutes: coerceMutes(data?.mutes),
    reads: coerceReads(data?.reads),
    pubAccess: data?.pubAccess && typeof data.pubAccess === 'object' ? data.pubAccess : {},
    dms: coerceDms(data?.dms),
    quickReactions: coerceQuickReactions(data?.quickReactions),
    archivedDms: coerceArchivedDms(data?.archivedDms),
    hash: res?.hash ?? null,
  };
}

export async function readSpaces(client: StarfishClient, userId: string): Promise<SpacesDoc> {
  try {
    return await pullSpacesDoc(client, userId);
  } catch (err) {
    console.error('[readSpaces] failed to pull spaces registry', err);
    return {
      spaces: [],
      caps: {},
      mutes: coerceMutes(undefined),
      reads: coerceReads(undefined),
      pubAccess: {},
      dms: {},
      quickReactions: [],
      archivedDms: {},
      hash: null,
    };
  }
}

export async function updateSpacesDoc(
  client: StarfishClient,
  userId: string,
  mutator: (cur: { spaces: Space[]; caps: CapMap; pubAccess: PubAccessMap }) => { spaces: Space[]; caps: CapMap; pubAccess: PubAccessMap },
): Promise<void> {
  const MAX_ATTEMPTS = 3;
  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    const { spaces, caps, mutes, reads, pubAccess, dms, quickReactions, archivedDms, hash } = await pullSpacesDoc(client, userId);
    const cur = { spaces, caps, pubAccess };
    const next = mutator(cur);
    if (next === cur) return;
    try {
      await client.push(
        spacesPush(userId),
        { v: 1, spaces: next.spaces, caps: next.caps, mutes, reads, pubAccess: next.pubAccess, dms, quickReactions, archivedDms },
        hash,
      );
      return;
    } catch (err) {
      if (err instanceof ConflictError && attempt < MAX_ATTEMPTS - 1) continue;
      throw err;
    }
  }
}

export async function updateMutesDoc(
  client: StarfishClient,
  userId: string,
  mutator: (cur: MutePrefs) => MutePrefs | null,
): Promise<void> {
  const MAX_ATTEMPTS = 3;
  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    const { spaces, caps, mutes, reads, pubAccess, dms, quickReactions, archivedDms, hash } = await pullSpacesDoc(client, userId);
    const next = mutator(mutes);
    if (!next) return;
    try {
      await client.push(spacesPush(userId), { v: 1, spaces, caps, mutes: next, reads, pubAccess, dms, quickReactions, archivedDms }, hash);
      return;
    } catch (err) {
      if (err instanceof ConflictError && attempt < MAX_ATTEMPTS - 1) continue;
      throw err;
    }
  }
}

export async function updateReadsDoc(
  client: StarfishClient,
  userId: string,
  mutator: (cur: ReadPrefs) => ReadPrefs | null,
): Promise<void> {
  const MAX_ATTEMPTS = 3;
  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    const { spaces, caps, mutes, reads, pubAccess, dms, quickReactions, archivedDms, hash } = await pullSpacesDoc(client, userId);
    const next = mutator(reads);
    if (!next) return;
    try {
      await client.push(spacesPush(userId), { v: 1, spaces, caps, mutes, reads: next, pubAccess, dms, quickReactions, archivedDms }, hash);
      return;
    } catch (err) {
      if (err instanceof ConflictError && attempt < MAX_ATTEMPTS - 1) continue;
      throw err;
    }
  }
}

export async function updateDmsDoc(
  client: StarfishClient,
  userId: string,
  mutator: (cur: DmMap) => DmMap | null,
): Promise<void> {
  const MAX_ATTEMPTS = 3;
  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    const { spaces, caps, mutes, reads, pubAccess, dms, quickReactions, archivedDms, hash } = await pullSpacesDoc(client, userId);
    const next = mutator(dms);
    if (!next) return;
    try {
      await client.push(spacesPush(userId), { v: 1, spaces, caps, mutes, reads, pubAccess, dms: next, quickReactions, archivedDms }, hash);
      return;
    } catch (err) {
      if (err instanceof ConflictError && attempt < MAX_ATTEMPTS - 1) continue;
      throw err;
    }
  }
}

export async function updateQuickReactionsDoc(
  client: StarfishClient,
  userId: string,
  mutator: (cur: string[]) => string[] | null,
): Promise<void> {
  const MAX_ATTEMPTS = 3;
  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    const { spaces, caps, mutes, reads, pubAccess, dms, quickReactions, archivedDms, hash } = await pullSpacesDoc(client, userId);
    const next = mutator(quickReactions);
    if (!next) return;
    try {
      await client.push(spacesPush(userId), { v: 1, spaces, caps, mutes, reads, pubAccess, dms, quickReactions: next, archivedDms }, hash);
      return;
    } catch (err) {
      if (err instanceof ConflictError && attempt < MAX_ATTEMPTS - 1) continue;
      throw err;
    }
  }
}

export async function updateArchivedDmsDoc(
  client: StarfishClient,
  userId: string,
  mutator: (cur: ArchivedDms) => ArchivedDms | null,
): Promise<void> {
  const MAX_ATTEMPTS = 3;
  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    const { spaces, caps, mutes, reads, pubAccess, dms, quickReactions, archivedDms, hash } = await pullSpacesDoc(client, userId);
    const next = mutator(archivedDms);
    if (!next) return;
    try {
      await client.push(spacesPush(userId), { v: 1, spaces, caps, mutes, reads, pubAccess, dms, quickReactions, archivedDms: next }, hash);
      return;
    } catch (err) {
      if (err instanceof ConflictError && attempt < MAX_ATTEMPTS - 1) continue;
      throw err;
    }
  }
}

export async function setDmMapping(
  client: StarfishClient,
  userId: string,
  peerUserId: string,
  spaceId: string,
): Promise<void> {
  await updateDmsDoc(client, userId, (cur) => (cur[peerUserId] === spaceId ? null : { ...cur, [peerUserId]: spaceId }));
}

export async function writeSpaces(
  client: StarfishClient,
  userId: string,
  spaces: Space[],
  _hash: string | null,
): Promise<void> {
  await updateSpacesDoc(client, userId, (cur) => ({ spaces, caps: cur.caps, pubAccess: cur.pubAccess }));
}

export async function reorderSpaces(client: StarfishClient, userId: string, order: string[]): Promise<void> {
  await updateSpacesDoc(client, userId, (cur) => {
    const byId = new Map(cur.spaces.map((s) => [s.id, s]));
    const next: Space[] = [];
    for (const id of order) {
      const s = byId.get(id);
      if (s) { next.push(s); byId.delete(id); }
    }
    for (const s of cur.spaces) if (byId.has(s.id)) next.push(s);
    const unchanged = next.length === cur.spaces.length && next.every((s, i) => s === cur.spaces[i]);
    if (unchanged) return cur;
    return { spaces: next, caps: cur.caps, pubAccess: cur.pubAccess };
  });
}

function newSpaceId(): string {
  return `sp-${randomId()}`;
}

export async function readSpaceAccess(
  client: StarfishClient,
  spaceId: string,
): Promise<{ owner: string | null; members: string[]; visibility: SpaceVisibility | null; name: string | null; image: string | null; hash: string | null }> {
  const res = await client.pull(spaceAccessPull(spaceId)).catch((err: unknown) => {
    if (err instanceof StarfishHttpError && err.status === 404) return null;
    throw err;
  });
  const data = res?.data as { owner?: string; members?: unknown[]; visibility?: string; name?: string; image?: string } | undefined;
  return {
    owner: typeof data?.owner === 'string' ? data.owner : null,
    members: Array.isArray(data?.members) ? data!.members!.filter((m): m is string => typeof m === 'string') : [],
    visibility: data?.visibility === 'public' ? 'public' : null,
    name: typeof data?.name === 'string' ? data.name : null,
    image: typeof data?.image === 'string' ? data.image : null,
    hash: res?.hash ?? null,
  };
}

export async function writeSpaceAccess(
  client: StarfishClient,
  spaceId: string,
  owner: string,
  members: string[],
  hash: string | null,
  meta?: SpaceMeta,
): Promise<void> {
  const name = meta?.name?.trim() || undefined;
  const image = meta?.image || undefined;
  const visibility = meta?.visibility === 'public' ? 'public' : undefined;
  await client.push(
    spaceAccessPush(spaceId),
    { v: 1, owner, members, ...(visibility ? { visibility } : {}), ...(name ? { name } : {}), ...(image ? { image } : {}) },
    hash,
  );
}

export async function addSpaceMember(
  client: StarfishClient,
  spaceId: string,
  ownerUserId: string,
  memberUserId: string,
): Promise<void> {
  const { owner, members, visibility, name, image, hash } = await readSpaceAccess(client, spaceId);
  if (memberUserId === (owner ?? ownerUserId) || members.includes(memberUserId)) return;
  await writeSpaceAccess(client, spaceId, owner ?? ownerUserId, [...members, memberUserId], hash, { name, image, visibility: visibility ?? undefined });
}

/** Remove a member from the space roster (used for link revocation). */
export async function removeSpaceMember(
  client: StarfishClient,
  spaceId: string,
  memberUserId: string,
): Promise<void> {
  const { owner, members, visibility, name, image, hash } = await readSpaceAccess(client, spaceId);
  if (!members.includes(memberUserId)) return;
  await writeSpaceAccess(client, spaceId, owner ?? memberUserId, members.filter((m) => m !== memberUserId), hash, { name, image, visibility: visibility ?? undefined });
}

export async function addJoinedSpace(client: StarfishClient, userId: string, space: Space): Promise<void> {
  await updateSpacesDoc(client, userId, (cur) =>
    cur.spaces.some((s) => s.id === space.id)
      ? cur
      : { spaces: [...cur.spaces, space], caps: cur.caps, pubAccess: cur.pubAccess },
  );
}

export async function addJoinedSpaceWithCap(
  client: StarfishClient,
  userId: string,
  space: Space,
  capJson: string,
): Promise<void> {
  await updateSpacesDoc(client, userId, (cur) => ({
    spaces: cur.spaces.some((s) => s.id === space.id) ? cur.spaces : [...cur.spaces, space],
    caps: { ...cur.caps, [space.id]: capJson },
    pubAccess: cur.pubAccess,
  }));
}

export async function addJoinedSpaceWithLinkAccess(
  client: StarfishClient,
  userId: string,
  space: Space,
  sealed: SealedBlob,
): Promise<void> {
  await updateSpacesDoc(client, userId, (cur) => ({
    spaces: cur.spaces.some((s) => s.id === space.id) ? cur.spaces : [...cur.spaces, space],
    caps: cur.caps,
    pubAccess: { ...cur.pubAccess, [space.id]: sealed },
  }));
}

/**
 * Create a new space owned by the identity. Seeds an empty object index
 * (encrypted for private, plaintext for public). Apps populate the index
 * with their own object types after creation.
 *
 * `opts.visibility` defaults to `'private'`.
 */
export async function createSpace(
  session: Session,
  name: string,
  opts?: { visibility?: SpaceVisibility },
): Promise<Space> {
  const { accountClient, userId } = session;
  const { spaces, hash } = await readSpaces(accountClient, userId);
  const trimmed = name.trim() || 'New Space';
  const visibility = opts?.visibility ?? 'private';
  const id = newSpaceId();
  const space: Space = {
    id,
    name: trimmed,
    short: trimmed.slice(0, 2).toUpperCase(),
    members: 1,
    ...(visibility === 'public' ? { visibility: 'public', ownerId: userId, write: true } : {}),
  };
  await writeSpaceAccess(accountClient, id, userId, [], null, { name: trimmed, visibility: visibility === 'public' ? 'public' : undefined });
  await seedSpaceObjectIndex(session, id, [], { visibility });
  await writeSpaces(accountClient, userId, [...spaces, space], hash);
  return space;
}

export async function reconcileSpaceMeta(
  client: StarfishClient,
  userId: string,
  spaceId: string,
  shared: SpaceMeta,
  knownSpaces?: Space[],
): Promise<void> {
  const sharedName = typeof shared.name === 'string' && shared.name.trim() ? shared.name : null;
  const sharedImage = typeof shared.image === 'string' && shared.image ? shared.image : null;
  if (sharedName === null && sharedImage === null) return;
  const known = knownSpaces?.find((s) => s.id === spaceId);
  if (known) {
    const name = sharedName ?? known.name;
    const short = name.slice(0, 2).toUpperCase();
    const image = sharedImage ?? known.image;
    if (name === known.name && short === known.short && (image ?? null) === (known.image ?? null)) return;
  }
  const { spaces, hash } = await readSpaces(client, userId);
  const cur = spaces.find((s) => s.id === spaceId);
  if (!cur) return;
  const name = sharedName ?? cur.name;
  const image = sharedImage ?? cur.image;
  const short = name.slice(0, 2).toUpperCase();
  if (name === cur.name && short === cur.short && (image ?? null) === (cur.image ?? null)) return;
  const next = spaces.map((s) => (s.id === spaceId ? { ...s, name, short, image } : s));
  await writeSpaces(client, userId, next, hash);
  broadcastSpaceMeta(spaceId, { name, short, image });
}

File: packages/python/octospaces-sdk/octospaces_sdk/spaces/registry.py
"""Spaces registry — ``_spaces`` and ``_access`` RMW helpers.

Mirrors ``packages/ts/octospaces-sdk/src/spaces/registry.ts``.
"""

from __future__ import annotations

import json
from typing import Any, Callable, Optional, TypedDict

from octospaces_sdk.core.ids import random_id, slugify
from octospaces_sdk.core.types import ID, CapMap, Space
from octospaces_sdk.sync.identity import Session
from octospaces_sdk.sync.paths import space_access_pull, space_access_push, spaces_pull, spaces_push

_MAX_RETRIES = 3


class SpaceMeta(TypedDict, total=False):
    name: Optional[str]
    image: Optional[str]


class SpaceMetaUpdate(TypedDict):
    name: str
    short: str
    image: Optional[str]


# Module-level listener bus for SpaceMeta updates
_meta_listeners: list[Callable[[str, SpaceMetaUpdate], None]] = []


def on_space_meta(fn: Callable[[str, SpaceMetaUpdate], None]) -> Callable[[], None]:
    _meta_listeners.append(fn)
    return lambda: _meta_listeners.remove(fn)


def broadcast_space_meta(space_id: str, meta: SpaceMetaUpdate) -> None:
    for fn in list(_meta_listeners):
        fn(space_id, meta)


# ── Spaces doc ────────────────────────────────────────────────────────────────


async def read_spaces(client: Any, user_id: str) -> dict[str, Any]:
    path = spaces_pull(user_id)
    doc = await client.pull(path)
    return doc.get("data", {}) if doc else {}


async def _update_spaces_doc(
    client: Any,
    user_id: str,
    mutator: Callable[[dict[str, Any]], dict[str, Any]],
) -> None:
    path_r = spaces_pull(user_id)
    path_w = spaces_push(user_id)
    for _ in range(_MAX_RETRIES):
        doc = await client.pull(path_r)
        data = doc.get("data", {}) if doc else {}
        base_hash = doc.get("hash") if doc else None
        updated = mutator(data)
        try:
            await client.push(path_w, updated, base_hash=base_hash)
            return
        except Exception as exc:
            if "conflict" in str(exc).lower():
                continue
            raise
    raise RuntimeError("Failed to update _spaces doc after retries")


async def update_spaces_doc(client: Any, user_id: str, mutator: Callable[[dict[str, Any]], dict[str, Any]]) -> None:
    await _update_spaces_doc(client, user_id, mutator)


async def write_spaces(client: Any, user_id: str, spaces: list[Space], base_hash: Optional[str]) -> None:
    path = spaces_push(user_id)
    await client.push(path, {"spaces": spaces}, base_hash=base_hash)


async def reorder_spaces(client: Any, user_id: str, order: list[str]) -> None:
    async def mutate(data: dict[str, Any]) -> dict[str, Any]:
        spaces = data.get("spaces", [])
        by_id = {s["id"]: s for s in spaces}
        reordered = [by_id[sid] for sid in order if sid in by_id]
        return {**data, "spaces": reordered}
    await _update_spaces_doc(client, user_id, mutate)


# ── Space access (``_access``) ────────────────────────────────────────────────


async def read_space_access(client: Any, space_id: str) -> dict[str, Any]:
    doc = await client.pull(space_access_pull(space_id))
    data = doc.get("data", {}) if doc else {}
    return {
        "owner": data.get("owner"),
        "members": data.get("members", []),
        "name": data.get("name"),
        "image": data.get("image"),
        "hash": doc.get("hash") if doc else None,
    }


async def write_space_access(
    client: Any,
    space_id: str,
    owner: str,
    members: list[str],
    base_hash: Optional[str],
    meta: Optional[dict[str, Any]] = None,
) -> None:
    data: dict[str, Any] = {"owner": owner, "members": members}
    if meta:
        data.update(meta)
    await client.push(space_access_push(space_id), data, base_hash=base_hash)


async def add_space_member(client: Any, space_id: str, owner_user_id: str, member_user_id: str) -> None:
    current = await read_space_access(client, space_id)
    members = list(current.get("members", []))
    if member_user_id not in members:
        members.append(member_user_id)
    await write_space_access(
        client, space_id, current["owner"], members, current["hash"],
        meta={"name": current.get("name"), "image": current.get("image")},
    )


async def remove_space_member(client: Any, space_id: str, member_user_id: str) -> None:
    current = await read_space_access(client, space_id)
    members = [m for m in current.get("members", []) if m != member_user_id]
    await write_space_access(
        client, space_id, current["owner"], members, current["hash"],
        meta={"name": current.get("name"), "image": current.get("image")},
    )


# ── Space creation ────────────────────────────────────────────────────────────


async def create_space(session: Session, name: str) -> Space:
    """Create a new space: write _access, seed empty object index, register in _spaces."""
    from octospaces_sdk.spaces.object_index import seed_space_object_index
    from octospaces_sdk.sync.paths import OBJECT_COLLECTIONS, owner_scope

    space_id = f"sp-{random_id()}"
    short = slugify(name)
    space = Space(id=space_id, name=name, short=short, members=1)

    # Write _access
    await write_space_access(
        session.spaces_registry_client,
        space_id,
        owner=session.user_id,
        members=[session.user_id],
        base_hash=None,
        meta={"name": name, "image": None},
    )

    # Seed empty object index
    await seed_space_object_index(session, space_id)

    # Add to _spaces registry
    async def add_space(data: dict[str, Any]) -> dict[str, Any]:
        spaces = data.get("spaces", [])
        caps = data.get("caps", {})
        return {**data, "spaces": [*spaces, space], "caps": caps}

    await _update_spaces_doc(session.account_client, session.user_id, add_space)
    return space


async def add_joined_space(client: Any, user_id: str, space: Space) -> None:
    async def mutate(data: dict[str, Any]) -> dict[str, Any]:
        spaces = data.get("spaces", [])
        if not any(s["id"] == space["id"] for s in spaces):
            spaces = [*spaces, space]
        return {**data, "spaces": spaces}
    await _update_spaces_doc(client, user_id, mutate)


async def add_joined_space_with_cap(client: Any, user_id: str, space: Space, cap_json: str) -> None:
    async def mutate(data: dict[str, Any]) -> dict[str, Any]:
        spaces = data.get("spaces", [])
        if not any(s["id"] == space["id"] for s in spaces):
            spaces = [*spaces, space]
        caps = dict(data.get("caps", {}))
        caps[space["id"]] = cap_json
        return {**data, "spaces": spaces, "caps": caps}
    await _update_spaces_doc(client, user_id, mutate)


async def add_joined_space_with_link_access(client: Any, user_id: str, space: Space, sealed: dict[str, Any]) -> None:
    async def mutate(data: dict[str, Any]) -> dict[str, Any]:
        spaces = data.get("spaces", [])
        if not any(s["id"] == space["id"] for s in spaces):
            spaces = [*spaces, space]
        pub_access = dict(data.get("pubAccess", {}))
        pub_access[space["id"]] = sealed
        return {**data, "spaces": spaces, "pubAccess": pub_access}
    await _update_spaces_doc(client, user_id, mutate)
